### PR TITLE
Show roles in project access page based on customization

### PR DESCRIPTION
**Pull request by vikram-raj**
Fix: https://issues.redhat.com/browse/ODC-5448

**Description**
Show roles in project access page role based on customization

**Screenshots/gif**
image

Kapture 2021-04-16 at 19 58 18

**Tests coverage**
image

**Test setup**
- Build the backend
- Start backend bridge with ```./bin/bridge -project-access-cluster-roles '["monitoring-edit", "edit", "admin"]'```

Note: putting it work in progress as the backend is not ready.

**Review this commit https://github.com/openshift/console/pull/8672/commits/aa3fd975b2ac4e113cc9cfe900c7f4a336e6302a**

### CODE REVIEW

**Comment by serenamarie125** [comment]
@vikram-raj this is great!  Curious if you could add a image or animated gif on how the admin customizes this?

**Comment by vikram-raj** [comment]
> @vikram-raj this is great! Curious if you could add a image or animated gif on how the admin customizes this?

Added GIF in PR description.

**Comment by andrewballantyne** [comment]
@serenamarie125 if you're asking how does an Admin ADD the dropdown option? That's being handled over here: https://github.com/openshift/console/pull/8533

File: frontend/packages/dev-console/src/components/project-access/hooks.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+export type ClusterRoleKind = K8sResourceCommon & {
+  rules: { [key: string]: string[] }[];
+};
+
+export enum accessRoles {
+  admin = 'Admin',
+  edit = 'Edit',
+  view = 'View',
+}
```
Comment (jerolimov) [inline frontend/packages/dev-console/src/components/project-access/hooks.ts:13]: I think this is more a defaultAccessRoles not instead of an enum, or?

```suggestion
export const defaultAccessRoles = {
  admin: 'Admin',
  edit: 'Edit',
  view: 'View',
};

```

File: frontend/packages/dev-console/src/components/project-access/hooks.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+// for testing until backend support
+// eslint-disable-next-line dot-notation
+window.SERVER_FLAGS['availableClusterRoles'] = ['admin', 'view', 'edit', 'cluster-reader'];
+export const useProjectAccessRoles = () => {
+  // eslint-disable-next-line dot-notation
+  const availableClustersRoles = window.SERVER_FLAGS['availableClusterRoles'];
+  const [clusterRoles, clusterRolesLoaded] = useK8sWatchResource<ClusterRoleKind[]>({
+    kind: ClusterRoleModel.kind,
+    isList: true,
+  });
```
Comment (jerolimov) [inline frontend/packages/dev-console/src/components/project-access/hooks.ts:23]: Can we check the `availableClusterRoles` here to fetch the ClusterRoles only if necessary?

```suggestion
  const [clusterRoles, clusterRolesLoaded] = useK8sWatchResource<ClusterRoleKind[]>(availableClustersRoles ? {
    kind: ClusterRoleModel.kind,
    isList: true,
  } : {});
```

File: frontend/packages/dev-console/src/components/project-access/hooks.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+  admin = 'Admin',
+  edit = 'Edit',
+  view = 'View',
+}
+// for testing until backend support
+// eslint-disable-next-line dot-notation
+window.SERVER_FLAGS['availableClusterRoles'] = ['admin', 'view', 'edit', 'cluster-reader'];
+export const useProjectAccessRoles = () => {
+  // eslint-disable-next-line dot-notation
+  const availableClustersRoles = window.SERVER_FLAGS['availableClusterRoles'];
```
Comment (jerolimov) [inline frontend/packages/dev-console/src/components/project-access/hooks.ts:19]: Small typo, let us call this availableClusterRoles similar to the server flag.

```suggestion
  const availableClusterRoles = window.SERVER_FLAGS['availableClusterRoles'];
```

File: frontend/packages/dev-console/src/components/project-access/hooks.ts (hunk 2 above)
Comment (vikram-raj) [reply frontend/packages/dev-console/src/components/project-access/hooks.ts:23]: We can not pass an empty object to `useK8sWatchResource` as kind is the required field.

**Comment by gajanan-more** [comment]
Screenshot from 2021-04-26 19-39-00
Screenshot from 2021-04-26 19-38-34


Working as expected.
/label qe-approved

**Comment by vikram-raj** [comment]
/hold

Wait for backend [PR](https://github.com/openshift/console/pull/8502) to get merged.

**Comment by rottencandy** [comment]
Since the epic acceptance criteria of [ODC-4650](https://issues.redhat.com/browse/ODC-4650) have not changed and the epic has an `px-ack`, we approve this on behalf of PX.

/label px-approved

**Comment by andrewballantyne** [comment]
@vikram-raj Typescript errors are being reported.

```
ERROR in /go/src/github.com/openshift/console/frontend/packages/dev-console/src/components/project-access/__tests__/hooks.spec.ts
ERROR in /go/src/github.com/openshift/console/frontend/packages/dev-console/src/components/project-access/__tests__/hooks.spec.ts(18,25):
TS2339: Property 'projectAccessClusterRoles' does not exist on type '{ alertManagerBaseURL: string; authDisabled: boolean; basePath: string; branding: string; consoleVersion: string; customLogoURL: string; customProductName: string; documentationBaseURL: string; ... 26 more ...; consolePlugins: string[]; }'.
```

**Comment by vikram-raj** [comment]
> @vikram-raj Typescript errors are being reported.
> 
> ```
> ERROR in /go/src/github.com/openshift/console/frontend/packages/dev-console/src/components/project-access/__tests__/hooks.spec.ts
> ERROR in /go/src/github.com/openshift/console/frontend/packages/dev-console/src/components/project-access/__tests__/hooks.spec.ts(18,25):
> TS2339: Property 'projectAccessClusterRoles' does not exist on type '{ alertManagerBaseURL: string; authDisabled: boolean; basePath: string; branding: string; consoleVersion: string; customLogoURL: string; customProductName: string; documentationBaseURL: string; ... 26 more ...; consolePlugins: string[]; }'.
> ```

It is expected as the backend is not yet merged.

File: frontend/packages/dev-console/locales/en/devconsole.json [hunk 4]
```diff
@@ -434,8 +434,8 @@
   "Rate of transmitted packets dropped": "Rate of transmitted packets dropped",
   "Average Container bandwidth by Pod: received": "Average Container bandwidth by Pod: received",
   "Average Container bandwidth by Pod: transmitted": "Average Container bandwidth by Pod: transmitted",
-  "Project access allows you to add or remove a user&apos;s access to the project. More advanced management of role-based access control appear in <2>Roles</2> and <5>Role Bindings</5>. For more information, see the <8>role-based access control documentation</8> .": "Project access allows you to add or remove a user&apos;s access to the project. More advanced management of role-based access control appear in <2>Roles</2> and <5>Role Bindings</5>. For more information, see the <8>role-based access control documentation</8> .",
```
Comment (jerolimov) [inline frontend/packages/dev-console/locales/en/devconsole.json:437]: Are this change requested by someone or should we keep the old version? Also if we don't have a translation for this strings now, we should avoid unnecessary changes. If there is a reason for that I'm absolutely fine with that change!

File: frontend/packages/dev-console/src/components/project-access/ProjectAccess.tsx [hunk 5]
```diff
@@ -23,21 +23,27 @@ import {
...
 import { validationSchema } from './project-access-form-validation-utils';
 import ProjectAccessForm from './ProjectAccessForm';
-import { Verb, UserRoleBinding, Roles, roleBinding } from './project-access-form-utils-types';
+import { Verb, UserRoleBinding, roleBinding } from './project-access-form-utils-types';
 
 export interface ProjectAccessProps {
   formName: string;
   namespace: string;
   roleBindings?: { data: []; loaded: boolean; loadError: {} };
+  roles: Roles;
```
Comment (jerolimov) [inline frontend/packages/dev-console/src/components/project-access/ProjectAccess.tsx:32]: Based on my change in the hook, you need to change this to:

```suggestion
  roles: { data: Roles; loaded: boolean };
```

This makes it easier to check if the data are loaded, instead of checking that the data are empty or not.

File: frontend/packages/dev-console/src/components/project-access/ProjectAccess.tsx [hunk 6]
```diff
@@ -23,21 +23,27 @@ import {
...
-const ProjectAccess: React.FC<ProjectAccessProps> = ({ formName, namespace, roleBindings }) => {
+const ProjectAccess: React.FC<ProjectAccessProps> = ({
+  formName,
+  namespace,
+  roleBindings,
+  roles,
+}) => {
   const { t } = useTranslation();
-  if (!roleBindings.loaded && _.isEmpty(roleBindings.loadError)) {
+  if ((!roleBindings.loaded && _.isEmpty(roleBindings.loadError)) || _.isEmpty(roles)) {
```
Comment (jerolimov) [inline frontend/packages/dev-console/src/components/project-access/ProjectAccess.tsx:42]: With the changed hook and changed props above, you can check for the loaded state, instead of expecting that the data are not empty.

```suggestion
  if ((!roleBindings.loaded && _.isEmpty(roleBindings.loadError)) || !roles.loaded) {
```

File: frontend/packages/dev-console/src/components/project-access/ProjectAccess.tsx [hunk 7]
```diff
@@ -23,21 +23,27 @@ import {
...
     return <LoadingBox />;
   }
 
-  const filteredRoleBindings = filterRoleBindings(roleBindings.data, Roles);
+  const filteredRoleBindings = filterRoleBindings(roleBindings.data, roles);
```
Comment (jerolimov) [inline frontend/packages/dev-console/src/components/project-access/ProjectAccess.tsx:46]: IMHO, the filter function must not need to know the frontend details, so we can just provide the keys here (the ClusterRole names)

```suggestion
  const filteredRoleBindings = filterRoleBindings(roleBindings.data, Object.keys(roles.data));
```

File: frontend/packages/dev-console/src/components/project-access/ProjectAccess.tsx [hunk 8]
```diff
@@ -108,8 +114,9 @@ const ProjectAccess: React.FC<ProjectAccessProps> = ({ formName, namespace, role
     <>
       <PageHeading>
         <Trans t={t} ns="devconsole">
-          Project access allows you to add or remove a user&apos;s access to the project. More
-          advanced management of role-based access control appear in{' '}
+          {
+            "Project access allows you to add or remove a user's access to the project. More advanced management of role-based access control appear in "
+          }
```
Comment (jerolimov) [inline frontend/packages/dev-console/src/components/project-access/ProjectAccess.tsx:119]: Is this change requested by someone or can we keep the old version?

Also if we don't have a translation for this strings now, I would personally avoid unnecessary changes to translation.

If there is a reason for that I'm absolutely fine with that change.

File: frontend/packages/dev-console/src/components/project-access/ProjectAccess.tsx [hunk 9]
```diff
@@ -131,7 +138,7 @@ const ProjectAccess: React.FC<ProjectAccessProps> = ({ formName, namespace, role
             onReset={handleReset}
             validationSchema={validationSchema}
           >
-            {(formikProps) => <ProjectAccessForm {...formikProps} />}
+            {(formikProps) => <ProjectAccessForm {...formikProps} roles={roles} />}
```
Comment (jerolimov) [inline frontend/packages/dev-console/src/components/project-access/ProjectAccess.tsx:141]: Based on the changed hook signature you need extract the roles here:

```suggestion
            {(formikProps) => <ProjectAccessForm {...formikProps} roles={roles.data} />}
```

File: frontend/packages/dev-console/src/components/project-access/__tests__/ProjectAccess.spec.tsx [hunk 10]
```diff
@@ -25,6 +26,7 @@ describe('Project Access', () => {
         loaded: false,
         loadError: {},
       },
+      roles: defaultAccessRoles,
```
Comment (jerolimov) [inline frontend/packages/dev-console/src/components/project-access/__tests__/ProjectAccess.spec.tsx:29]: Again. Because my recommendation is to change the hook return value, you need to change this test as well:

```suggestion
      roles: {
        data: defaultAccessRoles,
        loaded: true,
      },
```

File: frontend/packages/dev-console/src/components/project-access/project-access-form-utils.ts [hunk 11]
```diff
@@ -5,7 +5,22 @@ import {
...
+  admin: 'Admin',
+  edit: 'Edit',
+  view: 'View',
+};
+
+export type Roles = {
+  [key: string]: string;
+};
+
+export const getAvailableAccessRoles = () => {
```
Comment (jerolimov) [inline frontend/packages/dev-console/src/components/project-access/project-access-form-utils.ts:18]: Personal preferences is to provide types here, esp. with this JSON.parse

```suggestion
export const getAvailableAccessRoles = (): string[] | undefined => {
```

File: frontend/packages/dev-console/src/components/project-access/project-access-form-utils.ts [hunk 12]
```diff
@@ -5,7 +5,22 @@ import {
...
+  if (!window.SERVER_FLAGS.projectAccessClusterRoles) return undefined;
+  return JSON.parse(window.SERVER_FLAGS.projectAccessClusterRoles);
+};
+
+export const filterRoleBindings = (roleBindings: RoleBinding[], roles: Roles): RoleBinding[] => {
   return _.filter(roleBindings, (user: RoleBinding) => _.keys(roles).includes(user.roleRef.name));
 };
```
Comment (jerolimov) [inline frontend/packages/dev-console/src/components/project-access/project-access-form-utils.ts:25]: Also a personal preference which you can also ignore.. The Roles type is a "frontend type" which is not required for this utils function. I would simplify this and change the signature to `clusterRoleNames: string[]` and extract this data if required in the calling component.

```suggestion
export const filterRoleBindings = (roleBindings: RoleBinding[], clusterRoleNames: string[]): RoleBinding[] => {
  return _.filter(roleBindings, (user: RoleBinding) => clusterRoleNames.includes(user.roleRef.name));
};
```

File: frontend/packages/dev-console/src/components/project-access/hooks.ts [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+  }
+  const projectAccessRoles = projectClusterRoles.reduce((acc, projectClusterRole) => {
+    const clusterRoleName = projectClusterRole.metadata.name;
+    acc[clusterRoleName] =
+      projectClusterRole.metadata?.annotations?.['console.openshift.io/display-name'] ||
+      clusterRoleName;
+    return acc;
+  }, {});
+  return projectAccessRoles;
+};
```
Comment (jerolimov) [inline frontend/packages/dev-console/src/components/project-access/hooks.ts:33]: As we discussed this on slack and because I had some strange issues on my cluster, and not on a shared cluster I have done many experiments and debugging sessions in your hook. At the end I found that we have an issue with our Dropdown which doesn't support when the data (items) changed after the first rendering. (We should analyse this issue as well...)

I could not fix this with the usage of `useK8sWatchResource`. But with switch to `useK8sWatchResources` the UI works fine for me. The pro and con is that this method requested not all ClusterRoles but all configured and customized ClusterRoles individually. If the cluster admin configures no "availableClusterRoles" no API request is required.

I returned an extra `loaded` parameter so that the UI can show the LoadingBox bases on this information, so that it is not required that the hook returns an empty object for this use case (and the current Dropdown issue).

Please take a look and change whatever you want. At least on my local cluster this works fine. The fallback to the clusterRoleName works also if the API fails for any reason. (For example there is not such a ClusterRole name.)

```tsx
import * as React from 'react';
import { useK8sWatchResources } from '@console/internal/components/utils/k8s-watch-hook';
import { ClusterRoleModel } from '@console/internal/models';
import { K8sResourceCommon, WatchK8sResource } from '@console/internal/module/k8s';
import { defaultAccessRoles, getAvailableAccessRoles, Roles } from './project-access-form-utils';

export type ClusterRoleKind = K8sResourceCommon & {
  rules: { [key: string]: string[] }[];
};

export const useProjectAccessRoles = (): { data: Roles; loaded: boolean } => {
  const availableClusterRoles = getAvailableAccessRoles();

  const watchedClusterRoles = React.useMemo<Record<string, WatchK8sResource>>(() => {
    if (!availableClusterRoles || !availableClusterRoles.length) {
      return {};
    }
    return availableClusterRoles?.reduce((acc, role) => {
      acc[role] = {
        kind: ClusterRoleModel.kind,
        name: role,
      };
      return acc;
    }, {} as Record<string, WatchK8sResource>);
  }, [availableClusterRoles]);

  const clusterRoles = useK8sWatchResources<Record<string, ClusterRoleKind>>(watchedClusterRoles);

  if (!availableClusterRoles || !availableClusterRoles.length) {
    return { data: defaultAccessRoles, loaded: true };
  }

  const mappedRoles = availableClusterRoles.reduce((acc, role) => {
    const clusterRole = clusterRoles[role];
    if (clusterRole && clusterRole.loaded && clusterRole.data) {
      const label =
        clusterRole.data.metadata?.annotations?.['console.openshift.io/display-name'] ||
        clusterRole.data.metadata.name ||
        role;
      acc[role] = label;
    } else {
      acc[role] = role;
    }
    return acc;
  }, {} as Roles);

  const allLoadedOrFailed = Object.values(clusterRoles).every(
    (clusterRole) => clusterRole.loaded || clusterRole.loadError,
  );

  return { data: mappedRoles, loaded: allLoadedOrFailed };
};
```

**Review by jerolimov — CHANGES_REQUESTED** [review]
@vikram-raj As written in slack, I understand now why this change doesn't work while testing it locally. My cluster has more then 250 ClusterRoles because it has multiple additional operators installed. And because `useK8sWatchResource` returns all of them, first 250, then 300+ in my case, your empty check doesn't work. As we also already discussed, an update to the Dropdown items removes the selected title from it. It would be cool to fix this in the Dropdown component. But because this is a common used component I would recommend to keep the workaround and render the table after we know which cluster roles are available.

I added some recommendations to your code. Please take a look. I'm sure the change looks bigger then it is. The biggest change is in the `useProjectAccessRoles` hook itself.

And I'm sure that you need to adjust some more unit tests bases on this change.

**Comment by jerolimov** [comment]
Epic has already been approved by docs and noted that documentation is part of the stretch goal for 4.8 release.

/label docs-approved

File: frontend/packages/dev-console/src/components/project-access/ProjectAccess.tsx (hunk 8 above)
Comment (vikram-raj) [reply frontend/packages/dev-console/src/components/project-access/ProjectAccess.tsx:119]: This change fixes the `missing i18n key` error on the project access page.

**Comment by jerolimov** [comment]
Reviewed code, tested it locally and works fine.
/lgtm

**Comment by jerolimov** [comment]
I verified this change now again on a cluster, together with PR #8502

Started the cluster via clusterbot `launch openshift/console#8502,openshift/console#8672`

Added a ClusterRole with and one without display-name annotation:

```yaml
kind: ClusterRole
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: test1
  annotations:
    console.openshift.io/display-name: Test number one
rules: []
---
kind: ClusterRole
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: test2
rules: []
```

Searched the `cluster` Console resource and configured the following customisation:


Console path to YAML editor: `/k8s/cluster/operator.openshift.io~v1~Console/cluster/yaml`


```diff
apiVersion: operator.openshift.io/v1
kind: Console
metadata:
  name: cluster
spec:
+ customization:
+   projectAccess:
+     availableClusterRoles:
+       - admin
+       - test1
+       - test2
+       - test3
```

Open the developer perspective project page, “Project access” tab:

Console path to this page: `/project-details/ns/default/access`

Verified that ClusterRoles are RoleBindings could be created, updated and deleted.

Installed OpenShift Pipelines and OpenShift Serverless operator.

Created new ClusterRoles test5 and test6 and verified the customisation with this entries as well as we had issues earlier if a cluster has more then 250 ClusterRoles.

**Review by jerolimov — APPROVED** [review]
